backend/coins/eth/account.go

```diff
@@ -1028,7 +992,75 @@ func (account *Account) MatchesAddress(address string) (bool, error) {
 // EnqueueUpdate enqueues an update for the account.
 func (account *Account) EnqueueUpdate() {
 	select {
-	case account.enqueueUpdateCh <- struct{}{}:
+	case account.enqueueUpdateCh <- account:
 	default:
 	}
 }
+
+// QueryBalance queries the coin clientfor the balance of the account.
+func (account *Account) QueryBalance() (*big.Int, error) {
+	var balance *big.Int
+	var err error
+	if account.coin.erc20Token != nil {
+		balance, err = account.coin.client.ERC20Balance(account.address.Address, account.coin.erc20Token)
+	} else {
+		balance, err = account.coin.client.Balance(context.TODO(), account.address.Address)
+	}
+	if err != nil {
+		return nil, errp.WithStack(err)
+	}
+	return balance, nil
+}
+
+// UpdateBalances updates the balances of the accounts in the provided slice.
+func UpdateBalances(accounts []*Account, etherScanClient *etherscan.EtherScan) error {
+	ethNonErc20Addresses := make([]ethcommon.Address, 0, len(accounts))
+	for _, account := range accounts {
+		if account.isClosed() {
+			continue
+		}
+		address, err := account.Address()
+		if err != nil {
+			return errp.WithStack(err)
+		}
+		if account.coin.erc20Token == nil {
+			ethNonErc20Addresses = append(ethNonErc20Addresses, address.Address)
+		}
+	}
+
+	balances, err := etherScanClient.Balances(context.TODO(), ethNonErc20Addresses)
+	if err != nil {
+		return errp.WithStack(err)
+	}
+
+	for _, account := range accounts {
+		if account.isClosed() {
+			continue
+		}
+		address, err := account.Address()
```

Comment:
If an error happens anywhere here, we probably want to set the account state into critical error or set it to offline, and continue with the other accounts. Otherwise one account's error will prevent all subsequent ones from being updated.